src/Renderer/LayeredMaterial.js

```diff
@@ -41,288 +43,402 @@ var getColorAtIdUv = function(nbTex) {
+                moveElementsArray(this.uniforms.dTextures_01.value,oldOffset,texturesCount,offsetTexture);
+            }
 
-};
+            this.setTextureOffsetByLayerIndex(newIndex,offsetTexture);
+            offsetTexture += texturesCount;
+
+        } else {
+            offsetLayer++;
+        }
+    }
 
-LayeredMaterial.prototype.removeLayerColor = function(idLayer) {
-    var layerIndex = this.layerIdToIndex[idLayer];
-    if (!layerIndex) return;
+    this.uniforms.colorLayersCount.value = this.getColorLayersCount();
+};
 
-    var startIdTexture = this.paramLayers[layerIndex].x;
-    var nbTextures = this.getNbColorTexturesLayer(layerIndex);
+LayeredMaterial.prototype.removeColorLayer = function(layer) {
 
-    this.paramLayers.splice(layerIndex, 1);
-    this.paramBLayers.splice(layerIndex, 1);
-    this.paramLayers.push(vector4);
-    this.paramBLayers.push(vector2);
+    let layerIndex = this.indexOfColorLayer(layer);
 
-    for (var i = startIdTexture, max = startIdTexture + nbTextures; i < max; i++) {
-        if (this.Textures[1][i] instanceof THREE.Texture)
-            this.Textures[1][i].dispose();
+    if (layerIndex === -1) {
+        return;
     }
 
-    this.Textures[1].splice(startIdTexture, nbTextures);
-    this.uniforms.nColorLayer.value--;
+    var offset = this.getTextureOffsetByLayerIndex(layerIndex);
+    let texturesCount = this.getTextureCountByLayerIndex(layerIndex);
+
+    // remove layer
+    this.colorLayersId.splice(layerIndex, 1);
+    this.uniforms.colorLayersCount.value = this.getColorLayersCount();
+
+    // remove nb textures
+    this.layerTexturesCount.splice(layerIndex, 1);
+    this.layerTexturesCount.push(0);
+
+    // Remove Layers Parameters
+    this.uniforms.paramLayers.value.splice(layerIndex, 1);
+    this.uniforms.paramLayers.value.push(vector4);
+
+    // Remove visibility Parameters
+    this.uniforms.visibility.value.splice(layerIndex, 1);
+    this.uniforms.visibility.value.push(true);
 
-    for (var j = layerIndex, mx = this.paramLayers.length; j < mx; j++)
-        this.paramLayers[j].x -= nbTextures;
+    // Dispose Layers textures
+    for (let i = offset, max = offset + texturesCount; i < max; i++) {
+        if (this.textures[l_COLOR][i] instanceof THREE.Texture) {
+            this.textures[l_COLOR][i].dispose();
+        }
+    }
 
-    // Rebuild sequence
-    var sequence = this.uniforms.layerSequence.value;
-    var limit = false;
+    var removedTexturesLayer = this.textures[l_COLOR].splice(offset, texturesCount);
+    let loadedTexturesLayerCount = 0;
+    this.offsetScale[l_COLOR].splice(offset, texturesCount);
 
-    for (var l = 0; l < this.uniforms.nColorLayer.value; l++) {
-        if (limit || sequence[l] === layerIndex) {
-            limit = true;
-            sequence[l] = sequence[l + 1];
+    for (let i = removedTexturesLayer.length - 1; i >= 0; i--) {
+        if(-1 < removedTexturesLayer[i].level) {
+            loadedTexturesLayerCount++;
         }
+    }
```

Comment:
For loops like this one can be shortened using `reduce` and arrow functions.

Something like this (not tested):
```js
let loadedTexturesLayerCount = removedTexturesLayer.reduce((sum, texture) => sum + (-1 < texture.level), 0);
```